### ISSUE TITLE
Make "Reference" type generic?

### ISSUE THREAD

**Issue by preston** [microsoft/fhir-codegen#70]
Would it be possible to make the TypeScript version of Reference generic? It is very difficult to enforce constraints without this, and TypeScript supports complex "A | B | C | D"-style generics.

We are trying to strictly enforce all manner of constraints in our information models for both IG publishing and software development purposes, and such refinements would be very welcome!

**Comment by GinoCanessa** [comment]
I like the idea, but I am struggling with what the output should look like.

In `Bundle`, we use the generic to apply type information to `Bundle.entry.resource`, which actually contains a FHIR object.

However, the datatype [Reference](http://hl7.org/fhir/references.html#Reference) doesn't contain an object - it just has an optional `type` element (which points to the URI of the reference type).

Is the goal that `Reference<Patient>` forces the literal `Patient` into `Reference.type`?

Thoughts: @bkaney @jdjkelly @ir4y @jmandel ?

**Comment by preston** [comment]
Yup, exactly. `Reference` itself would be a generic type. Something along the lines of..
```
export interface Reference<T extends FhirResource> extends Element {
...
_type?: T;
...
}
``` 
..with the general pattern repeated for other resource types to match more closely with the FHIR specification and be more directly implementable without casting madness.

**Comment by GinoCanessa** [comment]
Updating here for visibility.  @jmandel suggested the following:

```
interface Reference<T extends Resource> {
 ...
  type?: `${T["resourceType"]}`,
 ...
}
```

Which, I believe is in line with what we are looking for (e.g., the `type` field is actually a string value).  I haven't tested it yet, but it looks like a step in the right direction.  I believe it will need a little more logic around `undefined` though.

**Comment by jdjkelly** [comment]
```
type?: `${T["resourceType"]}` as FhirResourceType
```

**Comment by jdjkelly** [comment]
I mean, if it's just the string....

```
interface Reference<T extends FhirResourceType> {
 ...
  type?: T
 ...
}
```

**Comment by ir4y** [comment]
It would be nice to check `Reference.reference`. 
```typescript
interface Reference<T extends Resource> {
 ...
  reference?: `${T["resourceType"]}/\s+`,
 ...
}
```
Now it is not possible because https://github.com/microsoft/TypeScript/issues/41160 has to be approved first.
Also, we need to handle local references somehow.

Regarding defining the `Reference.type` personally, I like the @jmandel  idea.
```typescript
interface Reference<T extends Resource> {
 ...
  type?: `${T["resourceType"]}`,
 ...
}
``` 

However, based on type definition https://www.hl7.org/fhir/references-definitions.html#Reference.type the valueset binding for `Reference.type` is extensible.
Accordingly this the most close definition to FHIR is
```typescript
interface Reference<ResourceType extends (FhirResourceType | string) = FhirResourceType> {
...
type: ResourceType;
...
};

let r1:Reference<"Patient">;
let r2:Reference<"AnotherValue">;
```

I don't this. ` | string` usage makes `Reference.type` weak and it will no catch typos.
However, it is how FHIR defines `Reference.type` and we have to support all FHIR valid cases.

 I would like to hear your thoughts.
 
 @GinoCanessa @jdjkelly @bkaney @preston

**Comment by bkaney** [comment]
Yes, this would be great @ir4y! I am looking forward to https://github.com/microsoft/TypeScript/issues/41160, but wonder how we could do this to be somewhat backward compatiable?
